torch/csrc/jit/passes/onnx.cpp

```diff
@@ -23,16 +23,19 @@ bool hasUsedHandle(Node *node) {
 
 } // anonymous namespace
 
-// Transform PythonOps and Cpp Ops into Node's that match ONNX semantics.
-// Argument aten indicates whether we should export ops as "ATen" ONNX ops if possible.
 void ToONNX(std::shared_ptr<tracer::TracingState>& state, bool aten) {
   // Check that the tracing state is live (it should be, because
   // you were supposed to request zero derivatives.)
   if (state->is_expired()) {
     throw std::logic_error("ToONNX: tracing state is expired");
   }
+  state->graph = ToONNXFromGraph(state->graph, aten);
+}
 
-  auto new_graph = std::make_shared<Graph>(state->graph->scope_root());
+// Transform PythonOps and Cpp Ops into Node's that match ONNX semantics.
+// Argument aten indicates whether we should export ops as "ATen" ONNX ops if possible.
+std::shared_ptr<Graph> ToONNXFromGraph(std::shared_ptr<Graph>& graph, bool aten) {
```

Comment:
Our normal idiom for passes is to mutate the graph in place, so we should do that here as well.
If you want to not mutate a graph, you can always copy before calling this function.

Reply:
This is the ONNX conversion so I think it makes sense to make a copy, as opposed to the other passes which indeed mutate in place

Reply:
I still think it makes sense to only copy it only if requested. For instance, you might first want to apply some optimizations that you know are useful to run before going to onnx. Then you would do

```
graph  = copy(orig)
// these optimizations are only a good idea if we are exporting, so we first copied the graph
my_before_onnx_pre_opt(graph)
to_onnx(graph)
```
If this pass copies the graph, you would have to make two copies.

Reply:
Actually, nevermind. The pass is written in a way that you _have_ to copy it for it to even work. You can disregard what I was saying.